File: node/src/edgeAuth.js
/**
 * Copyright 2019 Phenix Real Time Solutions, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const program = require('commander');
const TokenBuilder = require('./TokenBuilder');
const defaultExpirationInSeconds = 3600;

program
  .option('-y, --uri <uri>', 'The backend URI')
  .option('-u, --applicationId <applicationId>', 'The application ID')
  .option('-w, --secret <secret>', 'The application secret')
  .option('-l, --expiresInSeconds <timeInSeconds>', 'Token life time in seconds')
  .option('-e, --expiresAt <timestamp>', 'Token expires at timestamp measured in milliseconds since UNIX epoch')
  .option('-a, --authenticationOnly', 'Token can be used for authentication only')
  .option('-s, --streamingOnly', 'Token can be used for streaming only')
  .option('-p, --publishingOnly', 'Token can be used for publishing only')
  .option('-b, --capabilities <capabilities>', 'Comma separated list of capabilities, e.g. for publishing')
  .option('-z, --sessionId <sessionId>', 'Token is limited to the given session')
  .option('-x, --remoteAddress <remoteAddress>', 'Token is limited to the given remote address')
  .option('-o, --originStreamId <originStreamId>', '[STREAMING] Token is limited to the given origin stream')
  .option('-c, --channel <channelId>', '[STREAMING] Token is limited to the given channel')
  .option('-i, --channelAlias <channelAlias>', '[STREAMING] Token is limited to the given channel alias')
  .option('-m, --room <roomId>', '[STREAMING] Token is limited to the given room')
  .option('-n, --roomAlias <roomAlias>', '[STREAMING] Token is limited to the given room alias')
  .option('-t, --tag <tag>', '[STREAMING] Token is limited to the given origin stream tag')
  .option('-r, --applyTag <applyTag...>', '[REPORTING] Apply tag to the new stream');

program.parse(process.argv);

const options = program.opts();
const tokenBuilder = new TokenBuilder()
  .withApplicationId(options.applicationId)
  .withSecret(options.secret);

if (options.uri) {
  tokenBuilder.withUri(options.uri);
}

if (options.expiresAt !== undefined) {
  tokenBuilder.expiresAt(new Date(parseInt(options.expiresAt, 10)));
} else {
  tokenBuilder.expiresInSeconds(parseInt(options.expiresInSeconds || defaultExpirationInSeconds, 10));
}

if (options.authenticationOnly) {
  tokenBuilder.forAuthenticateOnly();
}

if (options.streamingOnly) {
  tokenBuilder.forStreamingOnly();
}

if (options.publishingOnly) {
  tokenBuilder.forPublishingOnly();
}

if (options.capabilities) {
  options.capabilities.split(',').forEach((capability) => tokenBuilder.withCapability(capability));
}

if (options.sessionId) {
  tokenBuilder.forSession(options.sessionId);
}

if (options.remoteAddress) {
  tokenBuilder.forRemoteAddress(options.remoteAddress);
}

if (options.originStreamId) {
  tokenBuilder.forOriginStream(options.originStreamId);
}

if (options.channel) {
  tokenBuilder.forChannel(options.channel);
}

if (options.channelAlias) {
  tokenBuilder.forChannelAlias(options.channelAlias);
}

if (options.room) {
  tokenBuilder.forRoom(options.room);
}

if (options.roomAlias) {
  tokenBuilder.forRoomAlias(options.roomAlias);
}

if (options.tag) {
  tokenBuilder.forTag(options.tag);
}

if (options.applyTag) {
  tokenBuilder.applyTags(options.applyTag);
}

const tokenObject = tokenBuilder.value();

console.log(tokenObject);

const token = tokenBuilder.build();

console.log(token);

File: python/src/edge_auth.py
import argparse
import readline

from pprint import PrettyPrinter
from sys import version_info

from token_builder import TokenBuilder

PY3_MIN_VERSION = (3, 6)

if version_info < PY3_MIN_VERSION:
    raise Exception('Please use Python >= 3.6')

parser = argparse.ArgumentParser(description='EdgeAuth token generator')


parser.add_argument('-u', '--application_id', required=True, help='The application ID')
parser.add_argument('-w', '--secret', required=True, help='The application secret')
parser.add_argument('-l', '--expires_in_seconds', type=int, default=3600, help='Token life time in seconds')
parser.add_argument('-e', '--expires_at', help='Token expires at timestamp measured in milliseconds since UNIX epoch')
parser.add_argument('-b', '--capabilities', help='Comma separated list of capabilities, e.g. for publishing')
parser.add_argument('-z', '--session_id', help='Token is limited to the given session')
parser.add_argument('-x', '--remote_address', help='Token is limited to the given remote address')

parser.add_argument('-o', '--origin_stream_id', help='[STREAMING] Token is limited to the given origin stream')
parser.add_argument('-c', '--channel', help='[STREAMING] Token is limited to the given channel')
parser.add_argument('-i', '--channel_alias', help='[STREAMING] Token is limited to the given channel alias')
parser.add_argument('-m', '--room', help='[STREAMING] Token is limited to the given room')
parser.add_argument('-n', '--room_alias', help='[STREAMING] Token is limited to the given room alias')
parser.add_argument('-t', '--tag', help='[STREAMING] Token is limited to the given origin stream tag')

parser.add_argument('-r', '--apply_tag', help='[REPORTING] Apply tag to the new stream')

parser.add_argument('-a', '--authentication_only', action="store_true", help='Token can be used for authentication only')
parser.add_argument('-s', '--streaming_only', action="store_true", help='Token can be used for streaming only')
parser.add_argument('-p', '--publishing_only', action="store_true", help='Token can be used for publishing only')


args = parser.parse_args()

pp = PrettyPrinter().pprint

token = TokenBuilder() \
    .with_application_id(args.application_id) \
    .with_secret(args.secret)


if args.expires_at is not None:
    token.expires_at(args.expires_at)
else:
    token.expires_in_seconds(args.expires_in_seconds)

pp(args)

pp(token.build())